### PR TITLE
chore: release docker image

### PR DESCRIPTION
**Pull request by mg98**
By creating a submodule in my repository that connects to your [docker-compose files](https://github.com/trudi-group/ipfs-metric-exporter/tree/master/docker-compose) **and** having your image in the cloud, I can very neatly and nicely decouple [my project](https://github.com/mg98/ipfs-replicate) from the building processes of yours and create a single docker-compose.yml that spins up the whole infrastructure.

To this end, I had to make some changes.

### Changes

- Your GH action will build and push an image to the Github Package Registry on every commit on master, which will then show on your repository
- Merged `Dockerfile.builder` and `Dockerfile.kubo` into a single one (same effect, no bloated image!)
- I let your `docker-compose.yml` build from `..`

### Notes

You should be able to safely merge it to master without further actions required.
- The image is <500MB so it's even covered by [Github's Free plan](https://docs.github.com/en/billing/managing-billing-for-github-packages/about-billing-for-github-packages).
- You don't need to set any repository secrets! Those are all magic and inherently available variables.
- **Maybe** you have to [allow read and write permissions for workflows](https://github.com/trudi-group/ipfs-metric-exporter/settings/actions).

I tested the workflow on my [fork's master branch](https://github.com/mg98/ipfs-metric-exporter).

### CODE REVIEW

File: .github/workflows/Lint_Build.yaml [hunk 1]
```diff
@@ -43,10 +43,6 @@ jobs:
           echo "${REVIVE_OUTPUT}"
           exit 1
         fi
-    - name: Build
-      run: go build -v -buildmode=plugin -o mexport.so
-    - name: Build Docker Image
```
Comment (mrd0ll4r) [inline .github/workflows/Lint_Build.yaml:48]: Will this still be built as part of the CI running for PRs?

File: build-in-docker.sh [hunk 2]
```diff
@@ -1,16 +0,0 @@
-#!/bin/bash -e
```
Comment (mrd0ll4r) [inline build-in-docker.sh:1]: I still need some script to copy out the two artifacts onto the host filesystem. Shouldn't be difficult with the merged image though...

File: README.md [hunk 3]
```diff
@@ -18,10 +18,7 @@ This gives us compatibility with slightly older systems (e.g. Ubuntu LTS release
 The [builder Dockerfile](./Dockerfile.builder) implements a builder stage.
```
Comment (mrd0ll4r) [inline README.md:18]: This is not a thing anymore

**Review by mrd0ll4r — CHANGES_REQUESTED** [review]
Nice, thanks. I made a few comments.
Also, the README in `docker-compose/` needs an update, since it still references the build script.

File: .github/workflows/Lint_Build.yaml (hunk 1 above)
Comment (mg98) [reply .github/workflows/Lint_Build.yaml:48]: No. Do you prefer it to run a test build?

File: build-in-docker.sh (hunk 2 above)
Comment (mg98) [reply build-in-docker.sh:1]: Don't understand. The artifacts equate to the new docker image, don't they?

Comment (mrd0ll4r) [reply build-in-docker.sh:1]: I don't only run this within docker. Sometimes I need to run it directly on the host for some testing, and then I need them extracted :)

File: Makefile [hunk 4]
```diff
@@ -31,6 +31,14 @@ mexport.so: main.go go.mod
 	$(GOCC) build $(GOFLAGS) -buildmode=plugin -o "$@" "$<"
 	chmod +x "$@"
 
+gen-out:
+	mkdir -p out
+	docker create --name extract $(docker build .)
+	docker cp extract:/usr/local/bin/ipfs ./out/
+	docker rm extract
+	mv out/ipfs/* out
```
Comment (mrd0ll4r) [inline Makefile:39]: The script should have changed: `/usr/local/bin/ipfs` in the container is now a single binary, where it used to be a directory containing the kubo binary and the plugin.

Comment (mg98) [reply Makefile:39]: Are you sure, it still gives me
```
$ ls out
ipfs-v0.17.0-docker  mexport-v0.17.0-docker.so
```
which I guess are IPFS + the plugin?

File: build-in-docker.sh (hunk 2 above)
Comment (mg98) [reply build-in-docker.sh:1]: Moved that to Makefile, as you saw already. Hope that's fine.

File: Makefile [hunk 5]
```diff
@@ -31,6 +31,12 @@ mexport.so: main.go go.mod
 	$(GOCC) build $(GOFLAGS) -buildmode=plugin -o "$@" "$<"
 	chmod +x "$@"
 
+gen-out:
+	mkdir -p out
+	docker create --name extract kubo-mexport
+	docker cp extract:/usr/local/bin/ipfs ./out/
```
Comment (mrd0ll4r) [inline Makefile:37]: This is still a single binary and will thus not extract the plugin:
```
Successfully tagged kubo-mexport:latest
leo@leo-thinkpad:~/tmp/ipfs-metric-exporter$ make gen-out
mkdir -p out
docker create --name extract kubo-mexport
db20b6b310959725af20edf86ca57ee77408f9638a3abddfbbfd4d221d7a8465
docker cp extract:/usr/local/bin/ipfs ./out/
docker rm extract
extract
leo@leo-thinkpad:~/tmp/ipfs-metric-exporter$ l out/
total 78M
drwxrwxr-x 2 leo leo 4.0K Jan 30 14:11 .
drwxrwxr-x 7 leo leo 4.0K Jan 30 14:11 ..
-rwxr-xr-x 1 leo leo  78M Dec 22 14:54 ipfs
```

File: README.md [hunk 6]
```diff
@@ -15,13 +15,7 @@ This is nice, because you get reproducible, matching binaries, compiled with Go
...
 This gives us compatibility with slightly older systems (e.g. Ubuntu LTS releases), at no loss of functionality.
 
-The [builder Dockerfile](./Dockerfile.builder) implements a builder stage.
-The resulting binaries are placed in `/usr/local/bin/ipfs/` inside the image.
-
-The [kubo Dockerfile](./Dockerfile.kubo) implements kubo with the bundled plugin.
-
-The [build-in-docker.sh](./build-in-docker.sh) script builds these images.
-It also and copies the produced binaries to the `out/` directory of the project.
+The [Dockerfile](./Dockerfile) implements kubo with the bundled plugin.
```
Comment (mrd0ll4r) [inline README.md:18]: Please add a sentence about the new make target :)

File: docker-compose/README.md [hunk 7]
```diff
@@ -10,9 +10,6 @@ This uses two images you'll need to build locally:
     This is most easily built using [the build script in the ipfs-tools repository](https://github.com/trudi-group/ipfs-tools/blob/master/build-docker-images.sh).
 - `monitoring-size-estimator:latest` is the size estimator.
     This, too, is most easily built using above build script.
-- `kubo-mexport:latest` is a recompiled kubo with our plugin.
```
Comment (mrd0ll4r) [inline docker-compose/README.md:13]: There's another reference to the `Dockerfile.kubo` further down in this README, which we should also change.

**Review by mrd0ll4r — CHANGES_REQUESTED** [review]

File: Makefile (hunk 5 above)
Comment (mg98) [reply Makefile:37]: Ohh I misunderstood what you said earlier. Got you. Working on it...

File: .gitignore [hunk 8]
```diff
@@ -1,2 +1 @@
-# Binaries produced by the build-in-docker.sh script.
-/out
+out/
```
[the remark this answers is not in the document]
Comment (mg98) [reply .gitignore:1]: You're right, I just verified it. Learned something today!

File: Makefile (hunk 5 above)
Comment (mg98) [reply Makefile:37]: Ok, we now get
```
# ls -l out
total 139632
-rwxr-xr-x 1 root root 81029040 Jan 12 16:46 ipfs
-rwxr-xr-x 1 root root 61950808 Jan 30 09:36 mexport-v0.17.0-docker.so
```

**Review by mrd0ll4r — APPROVED** [review]

**Pull request merged**